### ISSUE TITLE
keep events for timelines and trials separate

### ISSUE THREAD

**Issue by jodeleeuw** [jspsych/jsPsych#585]
e.g., allow `on_finish` and `on_start` for timelines that do not get inherited by trials.

**Comment by jodeleeuw** [comment]
I took a first pass at adding this feature on the `feature-timeline-events` branch.

I created `on_timeline_start` and `on_timeline_finish` as events.

A decision needs to be made about when these events should fire. Right now, I'm firing them when the `conditional` and `loop` functions would also fire. However, this revealed some odd behavior. 

For `on_timeline_start`, it fires:
* When the timeline begins in a simple timeline
* When the timeline resets due to the `repetitions` parameter being > 1
* **Not** on the repetitions due to when timeline variables are used.

For `on_timeline_finish` it fires:
* When the timeline ends in a simple timeline
* **Not** when the timeline resets due to `repetitions` being > 1
* **Not** on the repetitions due to when timeline variables are used.

Two questions:

1. What behavior makes sense here?
2. This probably means that the `conditional_function` fires on each repetition when `repetitions` is used. Is that desired?

**Comment by jodeleeuw** [comment]
This is now updated so that `on_timeline_finish` and `on_timeline_start` both run when the timeline is repeated due to `repetitions` being > 1. They do not run repeatedly when timeline variables are used.

I also fixed unintended behavior in the `conditional_function`, it was running multiple times when `repetitions` > 1. This is now changed so that the `conditional_function` executes at the start of the timeline, but not after. (Note that if a `loop_function` returns `true` then the timeline effectively starts over, so the `conditional_function` will run again.)

**Comment by jodeleeuw** [comment]
Handing this over to you now @becky-gilbert

**Comment by jodeleeuw** [comment]
I just discovered that `conditional_function` was running on every different entry in timeline variables.  I changed this so that it only runs with the first set of variables.

**Comment by jodeleeuw** [comment]
@becky-gilbert  I ended up merging the changes from this branch into the feature branch for #1493 because I needed some of the fixes from here to make the tests work in that branch. However, I think we still need docs for these changes so I left this branch open.

**Comment by becky-gilbert** [comment]
Got it, thanks. I'll do the docs and then close this out.

**Comment by becky-gilbert** [comment]
I'm on this `feature-timeilne-events` branch and these functions aren't working as I'd expected. The `on_timeline_start` function is called repeatedly with timeline variables, whereas the `on_timeline_finish` function isn't. Is that the intended behavior?

I tested this with new jest tests and by modifying the "timeline_variables.html" file:
```js
 var node = {
    timeline_variables: timeline_variables,
    timeline: [
      {
        type: 'html-keyboard-response',
        choices: jsPsych.NO_KEYS,
        stimulus: "<p style='text-align:center; font-size:80px;'>+</p>",
        trial_duration: 500
      },
      {
        type: 'image-keyboard-response',
        choices: ['y','n'], 
        stimulus: jsPsych.timelineVariable('v1'),
        prompt: function() { 
          return '<p>Have you seen '+jsPsych.timelineVariable('v2', true)+ ' before? Press "y" or "n".</p>' 
        },
        stimulus_width: 300
      }
    ],
    randomize_order: true,
    repetitions: 2,
    on_timeline_start: function() {
      console.log('timeline just started'); // <---- called at the start of every trial procedure (fixation)
    },
    on_timeline_finish: function() {
      console.log('timeline just finished');  // <----- called at the end of each repetition
    }
  }
```

**Comment by becky-gilbert** [comment]
I think we missed this problem with `on_timeline_start` and timeline variables because of an error in the tests for timeline variables. (The timeline variables had an array with just one object/trial, and I think it was supposed to have two.) I fixed this in the tests and now the test for `on_timeline_start` and timeline variables is failing. I'm happy to try to fix this but wanted to check whether this is the intended behavior first.

**Comment by becky-gilbert** [comment]
I'm going to merge this branch now that the problem with `on_timeline_start` and timeline variables is fixed, tests are passing, and I finished the docs.

**Comment by becky-gilbert** [comment]
Closed via #1512
